backend/x86_peephole/x86_peephole_rules.ml

```diff
@@ -0,0 +1,437 @@
+      (* Update cells with combined values *)
+      DLL.set_value cell1 (Ins (ADD (Imm combined_imm, Reg64 RSP)));
+      DLL.set_value cell2
+        (Directive
+           (Asm_targets.Asm_directives.make_cfi_adjust_cfa_offset_directive
+              combined_offset));
+      (* Delete the redundant cells *)
+      DLL.delete_curr cell3;
+      DLL.delete_curr cell4;
+      stats.combine_add_rsp <- stats.combine_add_rsp + 1;
+      (* Return cell1 to allow iterative combination of multiple ADDs *)
+      U.Matched (Some cell1)
+    | _, _, _, _ -> U.No_match)
+  | _ -> U.No_match
+
+(* Rewrite rule: optimize MOV chain that writes to intermediate register.
+   Pattern: mov A, x; mov x, y; mov B, x Rewrite: mov A, y; mov B, x
+
+   This is safe when B ≠ x (otherwise we'd incorrectly eliminate a write). The
+   transformation preserves the final values: x = B, y = A.
+
+   Additionally, both x and y must be registers to ensure the rewritten
+   instruction mov A, y is valid (x86 cannot have both operands as memory). *)
+let remove_mov_chain stats cell =
+  match U.get_cells cell 3 with
+  | [cell1; cell2; cell3] -> (
+    match DLL.value cell1, DLL.value cell2, DLL.value cell3 with
+    | Ins (MOV (src1, dst1)), Ins (MOV (src2, dst2)), Ins (MOV (src3, dst3))
+      when U.equal_args dst1 src2 && U.equal_args dst1 dst3
+           && (not (U.equal_args src3 dst3))
+           && U.is_register dst1 && U.is_register dst2 ->
+      (* Pattern: mov A, x; mov x, y; mov B, x where B ≠ x and x, y are
+         registers *)
+      (* Rewrite to: mov A, y; mov B, x *)
+      DLL.set_value cell1 (Ins (MOV (src1, dst2)));
+      DLL.set_value cell2 (Ins (MOV (src3, dst3)));
+      DLL.delete_curr cell3;
+      stats.remove_mov_chain <- stats.remove_mov_chain + 1;
+      (* Return cell1 to allow iterative combination of MOV chains *)
+      U.Matched (Some cell1)
+    | _, _, _ -> U.No_match)
+  | _ -> U.No_match
+
+(* Rewrite rule: optimize MOV to register that is overwritten before use.
+   Pattern: mov A, x; mov x, y where the next occurrence of x is a write.
+   Rewrite: mov A, y
+
+   This is safe when both x and y are registers and x is not read before the
+   next write to x within the same basic block. The transformation preserves
+   semantics: y gets the value of A, and x is overwritten before being read.
+
+   We restrict x to Reg64 to avoid register aliasing issues. *)
+let remove_mov_to_dead_register stats cell =
+  match U.get_cells cell 2 with
+  | [cell1; cell2] -> (
+    match DLL.value cell1, DLL.value cell2 with
+    | Ins (MOV (src1, dst1)), Ins (MOV (src2, dst2))
+      when U.equal_args dst1 src2 && U.is_register dst1 && U.is_register dst2
+           && U.is_safe_for_dead_register_opt dst1 -> (
+      (* Pattern: mov A, x; mov x, y where x and y are registers *)
+      (* Check if the next occurrence of x is a write *)
+      match U.find_next_occurrence_of_register dst1 cell2 with
+      | WriteFound ->
```

Comment:
`find_next_occurrence_of_register` returns `WriteFound` for writes to aliases, but don't we need the entire register to be overwritten here? I think it needs to distinguish between partial and full writes.